system/jlib/jstats.cpp

```diff
@@ -4011,6 +4051,277 @@ void verifyStatisticFunctions()
+        {
+            return std::hash<unsigned>{}((unsigned)sk);
+        }
+        bool operator()(const StatisticKind & lhs, const StatisticKind & rhs) const
+        {
+            return lhs==rhs;
+        }
+    };
+    ScopeId id;
+    std::unordered_map<StatisticKind, unsigned __int64, StatisticKindFunc, StatisticKindFunc> values;
+    std::unordered_map<ScopeId, Owned<StatsAggregator>, ScopeId, ScopeId> children;
+
+    StatsAggregator * queryAggregator(const ScopeId & scopeId)
+    {
+        if (auto search = children.find(scopeId); search != children.end())
+            return search->second;
+        return nullptr;
+    }
+    void eraseAggregator(const ScopeId & scopeId)
+    {
+        if (auto search = children.find(scopeId); search != children.end())
+            children.erase(search);
+    }
+public:
+    StatsAggregator() {}
+    StatsAggregator(ScopeId _id) : id(_id) {}
+    StatsAggregator * ensureScopeAggregator(const StatsScopeId & statScopeId)
+    {
+        ScopeId searchScope(statScopeId);
+        return ensureScopeAggregator(searchScope);
+    }
+    StatsAggregator * ensureScopeAggregator(const ScopeId & scopeId)
+    {
+        StatsAggregator * aggregator = queryAggregator(scopeId);
+        if (!aggregator)
+            aggregator = insertScope(scopeId);
+        return aggregator;
+    }
+    StatsAggregator * insertScope(const ScopeId & scopeId)
+    {
+        Owned<StatsAggregator> newScopeAggregator = new StatsAggregator(scopeId);
+        children.insert({scopeId, newScopeAggregator.getLink()});
+        return newScopeAggregator.getClear();
+    }
+    unsigned __int64 setValue(StatisticKind kind, unsigned __int64 value)
+    {
+        unsigned __int64 prevValue = values[kind];
+        values[kind] = value;
+        return value - prevValue;
+    }
+    void addValue(StatisticKind kind, unsigned __int64 value)
+    {
+        values[kind] += value;
+    }
+    void mergeInto(unsigned wfid, unsigned graphId, IStatisticGatherer & statisticGatherer, bool erase)
+    {
+        ScopeId wfScope(SSTworkflow, wfid);
+        StatsAggregator * wfScopeAggregator = queryAggregator(wfScope);
+        if (!wfScopeAggregator) // may happen if called before gatherStats called
+            return;
+
+        ScopeId graphScope(SSTgraph, graphId);
+        StatsAggregator * graphScopeAggregator = wfScopeAggregator->queryAggregator(graphScope);
+        if (!graphScopeAggregator) // may happen if called before gatherStats called
+            return;
+
+        graphScopeAggregator->mergeInto(statisticGatherer);
+        if (erase)
+           wfScopeAggregator->eraseAggregator(graphScope);
+
+    }
+    void mergeInto(IStatisticGatherer & statisticGatherer)
+    {
+        switch(id.queryScopeType())
+        {
+            case SSTglobal:
+            case SSTworkflow:
+            case SSTgraph:
+                // this IStatisticGatherer already at sg scope
```

Comment:
This looks like it is making use of knowledge of how the function is called - since line 4164 is beginning a scope.  The function should have consistent prerequisites for the different types.